### ISSUE TITLE
The separate player of video should disallow resizement to different aspect-ratios.

### ISSUE THREAD

**Issue by BEEDELLROKEJULIANLOCKHART** [telegramdesktop/tdesktop#24850]
### Steps to reproduce

1. Click any video.
2. Attempt to resize the player.

### Expected behaviour

Observe the title.

### Actual behaviour

The player is able to be resized to any aspect ratio. I know not why this is desirable, so resizement of the player should be limited to the aspect-ratio of the video that the player is playing.

![This](https://user-images.githubusercontent.com/42837531/182031469-da226b97-e4d7-4bc8-9d2f-399a7a680d6e.png) and ![this](https://user-images.githubusercontent.com/42837531/182031470-1d473653-f469-45ac-8e6c-22935d40b7ae.png) depict this problem.

<details>

If the problem is somehow specific to the original video, the original video is available at "https://t.me/UkraineWarPosts/2560".

</details>

### Operating system

What "https://download.fedoraproject.org/pub/fedora/linux/releases/36/Spins/x86_64/iso/Fedora-KDE-Live-x86_64-36-1.5.iso" installs:

```
[BEEDELLROKEJULIANLOCKHART@1656943212 ~]$ python3 -c 'import os; print(os.uname()); raise SystemExit'
posix.uname_result(sysname='Linux', nodename='1656943212', release='5.18.13-200.fc36.x86_64', version='#1 SMP PREEMPT_DYNAMIC Fri Jul 22 14:03:36 UTC 2022', machine='x86_64')
[BEEDELLROKEJULIANLOCKHART@1656943212 ~]$ uname -a
Linux 1656943212 5.18.13-200.fc36.x86_64 #1 SMP PREEMPT_DYNAMIC Fri Jul 22 14:03:36 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

### Version of Telegram Desktop

4.0.2

### Installation source

Other (unofficial) source

**Comment by ilya-fedin** [comment]
Unfortunately, the display server doesn't provide API to do that :(
On other systems this is done by changing window size&position, but Wayland doesn't provide any API to set position.

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
@ilya-fedin and @Aokromes, I'm not utilizing Wayland. I'm utilizing X.

**Comment by ilya-fedin** [comment]
@BEEDELLROKEJULIANLOCKHART no, you're using Wayland, this bug is a detector. You can ensure by checking "About System" category in system settings.

Creating issues on wayland repo is useless, there are already a few, wayland developers are hesitant at adding the required APIs. For example:
1. https://gitlab.freedesktop.org/wayland/wayland/-/issues/183
2. https://gitlab.freedesktop.org/wayland/wayland/-/issues/233
3. https://gitlab.freedesktop.org/wayland/wayland-protocols/-/issues/72

This is one of the places where Wayland sucks as a protocol due to its creators and nothing could be done about that. Only accepting and living with that (and bugs protocol design creates).

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
@ilya-fedin  Yeah, I hate Wayland too, especially that it is somehow touted as "the future". Consequently, I'm not using it:
```
[BEEDELLROKEJULIANLOCKHART@1656943212 ~]$ echo $XDG_SESSION_TYPE
x11
```

**Comment by ilya-fedin** [comment]
How does Telegram idenify your system in active session? It should say whether it uses X11 or Wayland. I believe Telegram is running via Wayland as this bug is known to happen only on Wayland.

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
I think that you'll need to provide to me a version that outputs that variable, because even
```
[BEEDELLROKEJULIANLOCKHART@1656943212 ~]$ xwininfo

xwininfo: Please select the window about which you
          would like information by clicking the
          mouse in that window.

xwininfo: Window id: 0x3600006 "Telegram (442540)"

  Absolute upper-left X:  1084
  Absolute upper-left Y:  372
  Relative upper-left X:  0
  Relative upper-left Y:  0
  Width: 1280
  Height: 1411
  Depth: 32
  Visual: 0x6af
  Visual Class: TrueColor
  Border width: 0
  Class: InputOutput
  Colormap: 0x3600005 (not installed)
  Bit Gravity State: NorthWestGravity
  Window Gravity State: NorthWestGravity
  Backing Store State: NotUseful
  Save Under State: no
  Map State: IsViewable
  Override Redirect State: no
  Corners:  +1084+372  -196+372  -196--343  +1084--343
  -geometry 1280x1411+1084+372

[BEEDELLROKEJULIANLOCKHART@1656943212 ~]$
```
does not appear to state what `telegram-desktop` uses.

**Comment by ilya-fedin** [comment]
> I expect that you'll need to provide to me a version that outputs `$XDG_SESSION_TYPE`, because even

This has nothing to do with XDG_SESSION_TYPE. It should send actual display server in use by Qt to the active session list.

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
@ilya-fedin, sorry: active session list? How should I acquire that for you?

**Comment by ilya-fedin** [comment]
Settings -> Privacy & Security -> Show all sessions
Click on thecurrent system, you will see system version. That's what you need to screenshot.

**Comment by Zamundaaa** [comment]
> Unfortunately, the display server doesn't provide API to do that :(
> On other systems this is done by changing window size&position, but Wayland doesn't provide any API to set position.

There is no need to set any positions in order to preserve the aspect ratio of a window

**Comment by ilya-fedin** [comment]
> There is no need to set any positions in order to preserve the aspect ratio of a window

How is that? :thinking:

**Comment by Zamundaaa** [comment]
You do not need to follow the mouse, the client is always free to resize how it wants.

> Or the resizing would feel weird.

Mpv does this, and it's not a problem unless really extreme aspect ratios are used.
If you think it's a real concern, you can render black bars around the video, like any other video player. Anything is better than messing up the aspect ratio

**Comment by ilya-fedin** [comment]
> You do not need to follow the mouse, the client is always free to resize how it wants.

If you would just resize without changing the position when the user is dragging left upper corner, it will expand to bottom right corner instead. It's a weird behavior.

> Mpv does this, and it's not a problem unless really extreme aspect ratios are used.

I have no idea whether mpv does this and how it does this. And I don't think I can find it in its codebase.

> If you think it's a real concern, you can render black bars around the video, like any other video player. Anything is better than messing up the aspect ratio

I'm not sure this is acceptable... I remember @john-preston raised concerns about impossibility to force window aspect ratio and I left it as is because it didn't move/resize at all otherwise.

**Comment by Zamundaaa** [comment]
It will not expand to the bottom right corner. The compositor handles it just like with any other interactive resize.

> And I don't think I can find this in its codebase

It's a really simple aspect ratio calculation: https://github.com/Dudemanguy/mpv/blob/4d3c71d2e2bde81e805ceb84b4e2eb072699327a/video/out/wayland_common.c#L997

**Comment by ilya-fedin** [comment]
> It will not expand to the bottom right corner. The compositor handles it just like with any other interactive resize.

I'm talking about changing the size _programmatically_. With interactive resize (aka system resize), there's no way to limit aspect ratio (that's what you currently see).

**Comment by ilya-fedin** [comment]
> It's a really simple aspect ratio calculation: https://github.com/Dudemanguy/mpv/blob/4d3c71d2e2bde81e805ceb84b4e2eb072699327a/video/out/wayland_common.c#L997

tdesktop also does calculation, it's not what is under interest. What's under interest is how it makes compositor to follow the aspect ratio.

**Comment by Zamundaaa** [comment]
> you mean QWindow::startSystemResize I guess

That is one way to start an interactive resize, the user dragging the server side decoration is the other (or using some keyboard shortcuts, or whatever). What I'm referring to is this: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/blob/main/unstable/xdg-shell/xdg-shell-unstable-v6.xml#L704
So while that resize is happening, clients can keep aspect ratio, resize their window to be only 10px high or whatever, and the compositor will adjust the position of the window automatically to fit that resize, no matter which edge the user is resizing from.

I can't tell you how to do this with Qt though, I don't know its API well enough for that.

**Comment by ilya-fedin** [comment]
> clients can keep aspect ratio

How? Thhe compositor tells new width and height (that won't follow aspect ratio of course as compositor doesn't know it) and the client can't do anything here.

**Comment by Zamundaaa** [comment]
The compositor tells the client the *maximum* width and height. You can make the window smaller in either or both directions, preserving the aspect ratio

**Comment by ilya-fedin** [comment]
> You can make the window smaller in either or both directions, preserving the aspect ratio

But if I'd call QWindow::setGeometry (that would call set_size in the protocol) while resizing, the resize process will be jagged, no? :thinking:

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
@ilya-fedin ![this](https://user-images.githubusercontent.com/42837531/182498100-ef75031d-2eac-4baa-8168-5bcc79279839.png) is response to "http://github.com/telegramdesktop/tdesktop/issues/24850#issuecomment-1203029807".

**Comment by ilya-fedin** [comment]
@BEEDELLROKEJULIANLOCKHART this is goes into "can't reproduce" state then as I can't reproduce this bug on X11.

**Comment by Zamundaaa** [comment]
> But if I'd call QWindow::setGeometry (that would call set_size in the protocol) while resizing, the resize process will be jagged, no? 

As long as you don't set multiple different geometries per resize request it should be fine. As this issue is apparently no longer about resize on Wayland, this conversation should maybe be taken elsewhere though?

**Comment by ilya-fedin** [comment]
> As long as you don't set multiple different geometries per resize request it should be fine.

I'm not quite sure what 'multiple different geometries per resize request' means, I tried to apply the following change:
```
diff --git a/Telegram/SourceFiles/media/view/media_view_pip.cpp b/Telegram/SourceFiles/media/view/media_view_pip.cpp
index faa564f25..b33322b82 100644
--- a/Telegram/SourceFiles/media/view/media_view_pip.cpp
+++ b/Telegram/SourceFiles/media/view/media_view_pip.cpp
@@ -572,6 +572,21 @@ void PipPanel::setGeometry(QRect geometry) {
 	widget()->setGeometry(geometry);
 }
 
+void PipPanel::handleResize(QSize size) {
+	if (!Platform::IsWayland()) {
+		return;
+	}
+
+	const auto max = std::max(size.width(), size.height());
+
+	// Apply aspect ratio.
+	const auto scaled = (_ratio.width() > _ratio.height())
+		? QSize(max, max * _ratio.height() / _ratio.width())
+		: QSize(max * _ratio.width() / _ratio.height(), max);
+
+	widget()->resize(scaled);
+}
+
 void PipPanel::handleMousePress(QPoint position, Qt::MouseButton button) {
 	if (button != Qt::LeftButton) {
 		return;
@@ -904,6 +919,11 @@ void Pip::setupPanel() {
 			break;
 		}
 	}, _panel.rp()->lifetime());
+
+	_panel.rp()->sizeValue(
+	) | rpl::start_with_next([=](QSize size) {
+		_panel.handleResize(size);
+	}, _panel.rp()->lifetime());
 }
 
 void Pip::handleClose() {
diff --git a/Telegram/SourceFiles/media/view/media_view_pip.h b/Telegram/SourceFiles/media/view/media_view_pip.h
index ec65ed206..0af3e9c22 100644
--- a/Telegram/SourceFiles/media/view/media_view_pip.h
+++ b/Telegram/SourceFiles/media/view/media_view_pip.h
@@ -76,6 +76,7 @@ public:
 	void setDragDisabled(bool disabled);
 	[[nodiscard]] bool dragging() const;
 
+	void handleResize(QSize size);
 	void handleMousePress(QPoint position, Qt::MouseButton button);
 	void handleMouseRelease(QPoint position, Qt::MouseButton button);
 	void handleMouseMove(QPoint position);
```
Result: resize is jagged and if you enable OpenGL rendering (default), resizing makes the window entirely black.

> As this issue is apparently no longer about resize on Wayland, this conversation should maybe be taken elsewhere though?

I have no idea for a better place. Only https://t.me/TelegramDesktopTalk maybe.

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
@ilya-fedin, please continue discussion at GitHub, because I am unable to easily track discussion of this issue elsewhere. If @Zamundaaa is confident that this should be discussed elsewhere, I suggest a pull-request.

**Comment by ilya-fedin** [comment]
Well, I remembered that someone told me about buffer size, so I hooked Qt with private APIs to override buffer size and this indeed works. This landed on dev today, so PiP on Wayland should follow aspect ratio with next version just like on X11. I don't think that will help you though as it's known to work just fine on X11 and it works on my machine (even with the video you provided). So either your system is broken or you're doing something special that you haven't mentioned in the reproduce steps.
